src/glang/lib/std/rb_tree.c3

```diff
@@ -0,0 +1,481 @@
+    }
+    return make_optional<K>();
+}
+
+function [K] minimum : (tree : RB_Tree<K>&) -> Optional<K> = {
+    return tree:minmax(LEFT());
+}
+
+function [K] maximum : (tree : RB_Tree<K>&) -> Optional<K> = {
+    return tree:minmax(RIGHT());
+}
+
+function [K] fix_after_delete_impl : (
+    self : RB_Tree<K>&, x : Optional<RB_Node<K>&>, y : RB_Node<K>&
+) -> Optional<RB_Node<K>&> = {
+    // x is guaranteed not to be the root. y is guaranteed to be x's parent.
+    let x_side : bool = which_child_am_i(x, &y);
+    let w_side : bool = !x_side;
+
+    // x's sibling, w, cannot be a leaf node (nil node). If x is double black,
+    // then its sibling tree must also have a height of at least 2 black nodes.
+    let w : RB_Node<K>& = &y:get_child(w_side):data();
+
+    // Case 1: w is red.
+    if w.colour == RED() {
+        // The rotation makes x's new sibling black.
+        self:rotate(&y, x_side);
+        y.colour = RED();
+        w.colour = BLACK();
+
+        // x is unchanged; we loop again and enter one of the other three cases.
+        return x;
+    }
+
+    // Case 2: w is black, w's children are black.
+    if w.left:is_black() and w.right:is_black() {
+        // Colour w red, to decrement the height of the sibling tree (in term of
+        // black nodes). This allows us to move x up to our parent: if y was
+        // black, then now it is double black; if it was red, then now it is
+        // red-and-black.
+        w.colour = RED();
+        return make_optional(&y);
+    }
+
+    // Case 3: w is black with red close and black distant child.
+    if w:get_child(x_side):is_red() and w:get_child(w_side):is_black() {
+        // Make w's red child w's new parent, and colour it black, such that x's
+        // sibling is now black. w must in turn become red, to compensate.
+        self:rotate(&w, w_side);
+        w.colour = RED();
+        w.parent:colour_black();
+
+        // x remains in place, and we get case 4.
+        return x;
+    }
+
+    // Case 4: w is black and w's distant child is red.
+    runtime_assert(w:get_child(w_side):is_red());
+
+    // Essentially, we make w y's parent, which means that there is an
+    // additional black node now in the path of x. This allows us to remove the
+    // double black. The root of the sibling tree is now w's red child, which
+    // we colour black to compensate.
+    self:rotate(&y, x_side);
+
+    // w takes y's colour, so the colour of this subtree's root is preserved.
+    w.colour = y.colour;
+
+    // We have removed one black node from w's side of the tree (w itself), so
+    // we need to make its red child black to compensate (as it is now the root
+    // of that subtree). On the other side, the new root should also be black,
+    // as it absorbs the "double black" from x.
+    w.left:colour_black();
+    w.right:colour_black();
+
+    // Remove the double black; we are done.
+    return self.root;
```

Comment:
Wow, the rest of it was quite easy but this took some thinking. Can confirm, this works in my head (and on my tablet).